text/0000-variant-types.md

```diff
@@ -0,0 +1,335 @@
+- Feature Name: variant_types
+- Start Date: 2016-01-07
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This is something of a two-part RFC, it proposes
+
+* making enum variants first-class types,
+* untagged enums (aka unions).
+
+The latter is part of the motivation for the former and relies on the former to
+be ergonomic.
+
+In the service of making variant types work, there is some digression into
+default type parameters for functions. However, that needs its own RFC to be
+spec'ed properly.
+
+
+# Motivation
+
+Enums are a convenient way of dealing with data which can be in one of many
+forms. When dealing with such data, it is typical to match, then perform some
+operations on the interior data. However, in many cases there is a large amount
+of processing to be done. Ideally we would factor that out into a function,
+passing the data to the function. However, currently in Rust, enum variants are
+not types and so we must choose an unsatisfactory work around - we pass
+each field of the variant separately (leading to unwieldy function signatures
+and poor maintainability), we pass the whole variant with enum type (and have to
+match again, with `unreachable!` arms in the function), or we embed a struct
+within the variant and pass the struct (duplicating data structures for no good
+reason). It would be much nicer if we could refer to the variant directly in the
+type system.
+
+When working with FFI code, we need to communicate with C programs which may use
+union data types. There is no way to represent a union in Rust, and thus working
+with such types is awkward and involves bug-prone transmutes. We should provide
+some way for Rust to handle such types.
+
+As we'll see below, variant types allow for an elegant solution to the union
+problem.
+
+
+# Detailed design - variant types
+
+Consider the example enum `Foo`:
+
+```rust
+pub enum Foo {
+    Variant1,
+    Variant2(i32, &'static str),
+    Variant3 { f1: i32, f2: &'static str },
+}
+```
+
+We create new instances by constructing one of the variants. The only type
+introduced is `Foo`. Variant names can only be used in patterns and for creating
+instances. E.g.,
+
+```rust
+fn new_foo() -> Foo {
+    Foo::Variant2(42, "Hello!")
+}
+```
+
+This RFC proposes allowing the programmer to use variant names as types, e.g.,
+
+```rust
+fn bar(x: Foo::Variant2) {}
+struct Baz {
+    field: Foo::Variant3,
+}
+```
+
+Both enums and their variants can currently be imported:
+
+```rust
+use Foo;
+use Foo::Variant1;
+```
+
+Importing an enum imports it into both the value and type namespace. Importing
+a variant imports it only into the value namespace. To maintain backwards
+compatibility, this will remain the default. In order to import an enum variant
```

Comment:
I can't check the code right now, but it looks like variants are already imported into both namespaces (and I do remember that they are *defined* in both namespaces):
```
use E::V;

enum E {
    V { field: u8 }
}

type V = u8; // Conflicts as expected

fn main() {
    let e = V{field: 0}; // Compiles as expected 
    match e {
        V{..} => {} // Compiles as expected 
    }
}
```